litmus-portal/authentication/api/routes/user_router.go

```diff
@@ -0,0 +1,184 @@
+	router.POST("/login", loginUser(service))
+	router.Use(middleware.JwtMiddleware())
+	router.POST("/update/password", updatePassword(service))
+	router.POST("/reset/password", resetPassword(service))
+	router.POST("/create", createUser(service))
+	router.POST("/update/details", updateUser(service))
+	router.GET("/users", fetchUsers(service))
+}
+
+func createUser(service user.Service) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		userRole := c.MustGet("role").(string)
+		if entities.Role(userRole) != entities.RoleAdmin {
+			c.AbortWithStatusJSON(utils.ErrorStatusCodes[utils.ErrUnauthorised], presenter.CreateErrorResponse(utils.ErrUnauthorised))
+			return
+		}
+		var userRequest entities.User
+		err := c.BindJSON(&userRequest)
+		if err != nil {
+			log.Warn(err)
+			c.JSON(utils.ErrorStatusCodes[utils.ErrInvalidRequest], presenter.CreateErrorResponse(utils.ErrInvalidRequest))
+			return
+		}
+		if userRequest.Role == "" || userRequest.UserName == "" || userRequest.Password == "" || userRequest.Email == "" {
+			c.JSON(utils.ErrorStatusCodes[utils.ErrInvalidRequest], presenter.CreateErrorResponse(utils.ErrInvalidRequest))
+			return
+		}
+		userRec, err := service.CreateUser(&userRequest)
+		if err == utils.ErrUserExists {
+			log.Info(err)
+			c.JSON(utils.ErrorStatusCodes[utils.ErrUserExists], presenter.CreateErrorResponse(utils.ErrUserExists))
+			return
+		}
+		if err != nil {
+			log.Error(err)
+			c.JSON(utils.ErrorStatusCodes[utils.ErrServerError], presenter.CreateErrorResponse(utils.ErrServerError))
+			return
+		}
+		c.JSON(200, userRec)
+	}
+}
+func updateUser(service user.Service) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		var userRequest entities.User
+		err := c.BindJSON(&userRequest)
+		if err != nil {
+			log.Warn(err)
+			c.JSON(utils.ErrorStatusCodes[utils.ErrInvalidRequest], presenter.CreateErrorResponse(utils.ErrInvalidRequest))
+			return
+		}
+		userRec, err := service.UpdateUser(&userRequest)
+		if err != nil {
+			c.JSON(utils.ErrorStatusCodes[utils.ErrServerError], presenter.CreateErrorResponse(utils.ErrServerError))
+		}
+		c.JSON(200, userRec)
+	}
+}
+func fetchUsers(service user.Service) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		users, err := service.GetUsers()
+		if err != nil {
+			log.Error(err)
+			c.JSON(utils.ErrorStatusCodes[utils.ErrServerError], presenter.CreateErrorResponse(utils.ErrServerError))
+			return
+		}
+		c.JSON(200, users)
+	}
+}
+
+func loginUser(service user.Service) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		var userRequest entities.User
+		err := c.BindJSON(&userRequest)
+		if err != nil {
+			log.Warn(err)
+			c.JSON(utils.ErrorStatusCodes[utils.ErrInvalidRequest], presenter.CreateErrorResponse(utils.ErrInvalidRequest))
+			return
+		}
+		if userRequest.UserName == "" || userRequest.Password == "" {
+			c.JSON(utils.ErrorStatusCodes[utils.ErrInvalidRequest], presenter.CreateErrorResponse(utils.ErrInvalidRequest))
+			return
+		}
```

Comment:
Better to sanitize the data before authenticating, so that trailing spaces could be trimmed

Reply:
The `litmus-frontend` already does this. As this a specific use-case API (only used by litmus-frontend), we can choose to avoid sanitisation of input data in favour to not add an additional dependency.

Reply:
I think there will be a problem if we remove trailing spaces from the password in the backend. Example- `pass  word` and `password` both will pass, if we remove trailing spaces.

Reply:
IMO we should atleast sanitise/validate username. Currently frontend is not handling this

Reply:
Yes, that can be done

Reply:
![image](https://user-images.githubusercontent.com/31009634/121293898-112dc400-c90a-11eb-9202-63c9d6d4d17b.png)

![image](https://user-images.githubusercontent.com/31009634/121293952-2c98cf00-c90a-11eb-94ff-6b388b07fa0b.png)



I assumed by validation, we are talking solely about trailing spaces, which are handled by the front-end as shown above. Is there any additional regex/username constraints we need to abide by?

Reply:
There should not be spaces at the end. "admin " should be trimmed to "admin"

Reply:
Ah, yes. We will need to add sanitisation. I was thinking should I create a middleware giving validation as well as sanitisation definitions for each route and pass it in the router. Something similar to [Joi](https://www.npmjs.com/package/joi) in the NodeJS world? Or would it be overkill for this project?

cc: @rajdas98 @gdsoumya @ishangupta-ds